openstax/middleware.py

```diff
@@ -38,4 +45,54 @@ def process_response(self, request, response):
             request.path_info = request.path
             response = self.handler.get_response(request)
 
-        return response
\ No newline at end of file
+        return response
+
+
+class RedirectMiddleware(MiddlewareMixin):
+    def process_response(self, request, response):
+        print('**[process_response] called')
+        # No need to check for a redirect for non-404 responses.
+        if response.status_code != 404:
+            return response
+
+        # Get the path
+        path = models.Redirect.normalise_path(request.get_full_path())
+        print('***path: ' + path)
+
+        # Find redirect
+        redirect = middleware.get_redirect(request, path)
+        if redirect is None:
+            # Get the path without the query string or params
+            path_without_query = urlparse(path).path
+
+            if path == path_without_query:
+                # don't try again if we know we will get the same response
+                return response
+
+            redirect = middleware.get_redirect(request, path_without_query)
+            if redirect is None:
+                return response
+
+        if redirect.link is None:
+            return response
+
+        if redirect.is_permanent:
+            self.send_google_analytics(request.get_host(),path)
+            return http.HttpResponsePermanentRedirect(redirect.link)
+        else:
+            self.send_google_analytics(request.get_host(),path)
+            return http.HttpResponseRedirect(redirect.link)
+
+    def send_google_analytics(self, host, path):
+        escaped_url = requests.utils.quote('https://' + host + path)
+        print('**[send_google_analytics] called: ' + escaped_url)
+        payload = {'v':'1','t':'pageView','cid':str(int(time() * 1000)),'tid':'UA-73668038-3','ec':'Redirects','ea':'clicked',
```

Comment:
Might be worth considering formatting this out with comments - this will be a bear to try and figure out if we have to change.
eg.
```
payload = {'v':'1', # API Version
           't':'pageView', # Hit type 
           'cid':str(int(time() * 1000)), # client id - using timestamp for uniqueness 
		   ...
```

Reply:
Done